File: 1823/FindWinnerOfCircularGame.cpp
/*
https://leetcode.com/problems/find-the-winner-of-the-circular-game/description/?envType=daily-question&envId=2024-07-08
*/

class Solution {
public:
    int findTheWinner(int n, int k) {
        int nextIdx=0;
        vector<int>all;
        for(int i=1; i<=n; ++i){
            all.push_back(i);
        }
        while(true){
            if(all.size()==1) break;
            all[(nextIdx+k-1)%all.size()]=-1;
            nextIdx=(nextIdx+k-1)%all.size();
            all.erase(remove(all.begin(), all.end(), -1), all.end());
        }
        return all[0];
    }
};

File: 1/1000s/18s/1823/FindWinnerOfCircularGame.py
# https://leetcode.com/problems/find-the-winner-of-the-circular-game/description/?envType=daily-question&envId=2024-07-08

class Solution:
    def findTheWinner(self, n: int, k: int) -> int:

        def removeFriendThatIsLost(currentListOfAllFriendsInGame, friendToRemove):
            return [friend for friend in currentListOfAllFriendsInGame if friend!=friendToRemove]

        indexToBegin=0
        currentListOfAllFriendsInGame=[]
        for i in range(1, n+1, 1):
            currentListOfAllFriendsInGame.append(i)
        while(1):
            if len(currentListOfAllFriendsInGame)==1:
                break
            currentListOfAllFriendsInGame[(indexToBegin+k-1)%len(currentListOfAllFriendsInGame)]=-1
            indexToBegin=(indexToBegin+k-1)%len(currentListOfAllFriendsInGame)
            currentListOfAllFriendsInGame=removeFriendThatIsLost(currentListOfAllFriendsInGame, -1)
        return currentListOfAllFriendsInGame[0]